Sources/HummingbirdFoundation/Files/FileIO.swift

```diff
@@ -102,105 +99,132 @@ public struct HBFileIO: Sendable {
+        switch contents {
+        case .byteBuffer(let buffer):
+            guard let buffer = buffer else { return }
+            try await self.writeFile(buffer: buffer, handle: handle, on: context.eventLoop)
+        case .stream(let streamer):
+            try await self.writeFile(stream: streamer, handle: handle, on: context.eventLoop)
         }
     }
 
     /// Load file as ByteBuffer
-    func loadFile(handle: NIOFileHandle, region: FileRegion, context: HBRequestContext) -> EventLoopFuture<HBResponseBody> {
-        return self.fileIO.read(
+    func loadFile(handle: NIOFileHandle, region: FileRegion, context: HBRequestContext) async throws -> HBResponseBody {
+        let buffer = try await self.fileIO.read(
             fileHandle: handle,
             fromOffset: Int64(region.readerIndex),
             byteCount: region.readableBytes,
             allocator: context.allocator,
             eventLoop: context.eventLoop
-        ).map { buffer in
-            return .byteBuffer(buffer)
-        }
+        ).get()
+        return .byteBuffer(buffer)
     }
 
     /// Return streamer that will load file
-    func streamFile(handle: NIOFileHandle, region: FileRegion, context: HBRequestContext) -> HBResponseBody {
-        let fileStreamer = FileStreamer(
-            handle: handle,
+    func streamFile(handle: NIOFileHandle, region: FileRegion, context: HBRequestContext) throws -> HBResponseBody {
+        let fileStreamer = try AsyncSequenceResponseBodyStreamer(FileAsyncSequence(
+            fileHandle: handle,
+            chunkSize: self.chunkSize,
             fileRegion: region,
             fileIO: self.fileIO,
-            chunkSize: self.chunkSize,
-            allocator: context.allocator
-        )
+            allocator: context.allocator,
+            eventLoop: context.eventLoop
+        ))
         return .stream(fileStreamer)
     }
 
     /// write byte buffer to file
-    func writeFile(buffer: ByteBuffer, handle: NIOFileHandle, on eventLoop: EventLoop) -> EventLoopFuture<Void> {
-        return self.fileIO.write(fileHandle: handle, buffer: buffer, eventLoop: eventLoop)
+    func writeFile(buffer: ByteBuffer, handle: NIOFileHandle, on eventLoop: EventLoop) async throws {
+        return try await self.fileIO.write(fileHandle: handle, buffer: buffer, eventLoop: eventLoop).get()
     }
 
     /// write output of streamer to file
-    func writeFile(stream: HBByteBufferStreamer, handle: NIOFileHandle, on eventLoop: EventLoop) -> EventLoopFuture<Void> {
-        return stream.consumeAll(on: eventLoop) { buffer in
-            return self.fileIO.write(fileHandle: handle, buffer: buffer, eventLoop: eventLoop)
+    func writeFile(stream: HBByteBufferStreamer, handle: NIOFileHandle, on eventLoop: EventLoop) async throws {
+        for try await buffer in stream.sequence {
+            try await self.fileIO.write(fileHandle: handle, buffer: buffer, eventLoop: eventLoop).get()
```

Comment:
Is this an `append` operation? I thought this (over)wrote the start of the buffer

Reply:
This is an append operation. If you open the file with append it'll append onto the end of it, but if you open the file with create then it will clear the contents of the file and then each subsequent write will append.

I can add a test to verify if you'd like

Reply:
I edited `testWriteLargeFile` to ensure it writes a streamed request